src/executor/query/TopNExecutor.cpp

```diff
@@ -0,0 +1,101 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "executor/query/TopNExecutor.h"
+#include "planner/Query.h"
+#include "util/ScopedTimer.h"
+
+
+namespace nebula {
+namespace graph {
+
+folly::Future<Status> TopNExecutor::execute() {
+    SCOPED_TIMER(&execTime_);
+    auto* topn = asNode<TopN>(node());
+    auto iter = ectx_->getResult(topn->inputVar()).iter();
+    if (UNLIKELY(iter == nullptr)) {
+        return Status::Error("Internal error: nullptr iterator in topn executor");
+    }
+    if (UNLIKELY(iter->isGetNeighborsIter())) {
+        std::string errMsg = "Internal error: TopN executor does not supported GetNeighborsIter";
+        LOG(ERROR) << errMsg;
+        return Status::Error(errMsg);
+    }
+    if (iter->isSequentialIter()) {
+        auto seqIter = static_cast<SequentialIter*>(iter.get());
+        auto &factors = topn->factors();
+        auto &colIndices = seqIter->getColIndices();
+        std::vector<std::pair<size_t, OrderFactor::OrderType>> indexes;
+        for (auto &factor : factors) {
+            auto indexFind = colIndices.find(factor.first);
+            if (indexFind == colIndices.end()) {
+                LOG(ERROR) << "Column name `" << factor.first
+                           << "' does not exist.";
+                return Status::Error("Column name `%s' does not exist.",
+                                     factor.first.c_str());
+            }
+            indexes.emplace_back(std::make_pair(indexFind->second, factor.second));
+        }
+        auto comparator = [&indexes] (const LogicalRow &lhs, const LogicalRow &rhs) {
+            for (auto &item : indexes) {
+                auto index = item.first;
+                auto orderType = item.second;
+                if (lhs[index] == rhs[index]) {
+                    continue;
+                }
+
+                if (orderType == OrderFactor::OrderType::ASCEND) {
+                    return lhs[index] < rhs[index];
+                } else if (orderType == OrderFactor::OrderType::DESCEND) {
+                    return lhs[index] > rhs[index];
+                }
+            }
+            return false;
+        };
+
+        auto offset = topn->offset();
+        auto count = topn->count();
+        auto size = seqIter->size();
+        int64_t maxCount = count;
+        int64_t heapSize = 0;
+        if (size <= static_cast<size_t>(offset)) {
+            maxCount = 0;
+        } else if (size > static_cast<size_t>(offset + count)) {
+            heapSize = offset + count;
```

Comment:
I'm confused about the heap size is not count.

Reply:
When the `offset` of Limit is 0, TopN executor gets the top `count` rows of data;
and when `offset` is not 0, TopN firstly gets the top `offset+count` rows of data, and then retrive `count` rows starting from `offset`. This is useful when offset is not 0 but closely to 0.
Because currently we can't know the amout of data in execution plan period, so the TopN optimization rule is just applied when offset is 0.

Reply:
So it's meaningful only in the situation that offset is 0. In my opinion, TopN should not contains `offset`, it's different from Limit here.

Reply:
I think topn=sort+limit.
When `offset` of `limit` is not 0 but closely to 0, eg. 1 or 2 or3, it may also useful to apply topn rule.
If we support CBO in the future, we may know the approximate amount of data in the optimization period so that we can evaluate whether it is necessary to apply topn rule.
For example, if data is about 10000 rows, and `offset` of `limit` is 3, it's obviously useful to apply topn rule, even though the `offset` is not 0. And if `offset` is 9990, it's obivously not usefule to apply topn rule.